### PR TITLE
TF-613: As a user, I wan to delete a filter email rule

### PR DESCRIPTION
**Pull request by ManhNTX**
https://user-images.githubusercontent.com/107173849/183816628-833d6d16-79d7-4094-874d-861febda05bc.mov

https://user-images.githubusercontent.com/107173849/183816872-00889d07-bb38-481b-99dd-dcf9bb810865.mov

### CODE REVIEW

File: lib/features/manage_account/data/datasource/manage_account_datasource.dart [hunk 1]
```diff
@@ -20,4 +20,7 @@ abstract class ManageAccountDataSource {
   Future<void> persistLanguage(Locale localeCurrent);
 
   Future<List<TMailRule>> getAllTMailRule(AccountId accountId);
+
+  Future<List<TMailRule>> deleteMailRule(AccountId accountId, List<TMailRule> listMailRule);
```
Comment (hoangdat) [inline lib/features/manage_account/data/datasource/manage_account_datasource.dart:24]: please more detail on this, `deleteRule` and `currentRuleList`

File: lib/features/manage_account/data/network/manage_account_api.dart [hunk 2]
```diff
@@ -142,4 +147,37 @@ class ManageAccountAPI {
 
     return result?.list.first.rules ?? <TMailRule>[];
   }
+
+  Future<List<TMailRule>> updateListTMailRule(
```
Comment (hoangdat) [inline lib/features/manage_account/data/network/manage_account_api.dart:151]: we can use the same API for edit, create, delete. But in `datasource` and `repository`, please define each operator separately

File: lib/features/manage_account/domain/model/delete_email_rule_request.dart [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
+import 'package:equatable/equatable.dart';
+import 'package:rule_filter/rule_filter/tmail_rule.dart';
+
+class DeleteEmailRuleRequest with EquatableMixin {
+
+  final int indexOfMailRuleDelete;
```
Comment (hoangdat) [inline lib/features/manage_account/domain/model/delete_email_rule_request.dart:7]: index is ok? why not object?

File: lib/features/manage_account/data/datasource/manage_account_datasource.dart [hunk 4]
```diff
@@ -20,4 +21,7 @@ abstract class ManageAccountDataSource {
   Future<void> persistLanguage(Locale localeCurrent);
 
   Future<List<TMailRule>> getAllTMailRule(AccountId accountId);
+
+  Future<List<TMailRule>> updateListTMailRule(AccountId accountId, DeleteEmailRuleRequest deleteEmailRuleRequest);
```
Comment (hoangdat) [inline lib/features/manage_account/data/datasource/manage_account_datasource.dart:25]: Please update the method name

File: lib/features/manage_account/presentation/email_rules/email_rules_controller.dart [hunk 5]
```diff
@@ -49,12 +59,16 @@ class EmailRulesController extends BaseController {
...
+        emailRuleDelete : emailRule,
+        listEmailRule: listEmailRule,
+    );
+    consumeState(_deleteEmailRuleInteractor.execute(_accountDashBoardController.accountId.value!, deleteEmailRuleRequest));
   }
 
-  void _getAllRules(AccountId accountId) {
-    consumeState(_getAllRulesInteractor.execute(accountId));
+  void _getAllRules() {
+    consumeState(_getAllRulesInteractor.execute(_accountDashBoardController.accountId.value!));
```
Comment (hoangdat) [inline lib/features/manage_account/presentation/email_rules/email_rules_controller.dart:71]: `!` is it safe?

File: lib/features/manage_account/domain/model/delete_email_rule_request.dart [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import 'package:equatable/equatable.dart';
+import 'package:rule_filter/rule_filter/tmail_rule.dart';
+
+class DeleteEmailRuleRequest with EquatableMixin {
+  final TMailRule emailRuleDelete;
+  final List<TMailRule> listEmailRule;
```
Comment (hoangdat) [inline lib/features/manage_account/domain/model/delete_email_rule_request.dart:6]: `currentEmailRules`

File: lib/features/manage_account/data/datasource_impl/manage_account_datasource_impl.dart [hunk 7]
```diff
@@ -71,4 +74,22 @@ class ManageAccountDataSourceImpl extends ManageAccountDataSource {
...
+  @override
+  Future<List<TMailRule>> updateListTMailRule(AccountId accountId, DeleteEmailRuleRequest deleteEmailRuleRequest) {
+
+    deleteEmailRuleRequest.listEmailRule.remove(deleteEmailRuleRequest.emailRuleDelete);
+
+    var mapIndexTMailRule = deleteEmailRuleRequest.listEmailRule.asMap();
+
+    mapIndexTMailRule = mapIndexTMailRule.map((key, value) => MapEntry(key, value.copyWith(id: RuleId(id: Id(key.toString())))));
+
+    final listTMailRule = mapIndexTMailRule.values.toList();
```
Comment (hoangdat) [inline lib/features/manage_account/data/datasource_impl/manage_account_datasource_impl.dart:87]: srr, but hard to understand

File: lib/features/manage_account/presentation/email_rules/email_rules_controller.dart (hunk 5 above)
Comment (ManhNTX) [reply lib/features/manage_account/presentation/email_rules/email_rules_controller.dart:71]: it safe

**Review by hoangdat — APPROVED** [review]

File: lib/features/manage_account/data/network/manage_account_api.dart [hunk 8]
```diff
@@ -122,9 +126,8 @@ class ManageAccountAPI {
     final processingInvocation = ProcessingInvocation();
     final requestBuilder = JmapRequestBuilder(_httpClient, processingInvocation);
 
-    final getRuleFilterMethod = GetRuleFilterMethod(
-      accountId,
-    );
+    final getRuleFilterMethod = GetRuleFilterMethod(accountId)
```
Comment (dab246) [inline lib/features/manage_account/data/network/manage_account_api.dart:129]: why change it?

File: lib/features/manage_account/data/network/manage_account_api.dart [hunk 9]
```diff
@@ -142,4 +145,42 @@ class ManageAccountAPI {
...
+
+    final setRuleFilterMethod = SetRuleFilterMethod(accountId)
+      ..addUpdateRuleFilter({Id(RuleFilterIdType.singleton.value): listTMailRuleWithId});
+
+    requestBuilder.invocation(setRuleFilterMethod);
+
+    final getListTMailRuleUpdated = GetRuleFilterMethod(accountId)
+      ..addIds({RuleFilterIdSingleton.ruleFilterIdSingleton.id});
+
+    final getMailboxUpdatedInvocation = requestBuilder.invocation(getListTMailRuleUpdated);
```
Comment (dab246) [inline lib/features/manage_account/data/network/manage_account_api.dart:168]: please rename it

File: lib/features/manage_account/data/network/manage_account_api.dart [hunk 10]
```diff
@@ -142,4 +145,42 @@ class ManageAccountAPI {
...
+
+    final response = await (requestBuilder
+      ..usings(getListTMailRuleUpdated.requiredCapabilities))
```
Comment (dab246) [inline lib/features/manage_account/data/network/manage_account_api.dart:171]: indent

**Review by dab246 — COMMENTED** [review]
Avoid formatting code too much

**Review by dab246 — APPROVED** [review]

**Pull request merged**